### PR TITLE
fix: Prevent relative install destinations leaking outside package install dir

### PR DESCRIPTION
**Pull request by gridbugs**
This disallows install dst paths from begining with ".." to prevent them referencing a directory outside the install directories of the package.

This new constraint would prevent globs from being able to refer to files outside the current directory and its descendants, so this also adds a new keyword which changes the prefix of paths matched by a glob in the install stanza. It's similar to the `(files (foo as bar))` syntax in that it allows installed files to be placed in a location other than the one implied by their position in the source tree. It's now possible to write:
`(files (glob_files (some/path/* with_prefix new/path)))` ...which takes all the paths matched by the glob and replaces the `some/path` component with `new/path`.

I haven't yet updated the docs or changelog or written any tests beside modifying the one defined in https://github.com/ocaml/dune/pull/8265 to demonstrate how the feature gets used to give us a chance to discuss and change things about this solution before taking the time to write docs/tests, etc.

Also note that this breaks the test `github2123`. I followed the rabbit hole starting from https://github.com/ocaml/dune/issues/2123 to understand why this is necessary but I still don't understand. The test uses this dune file:
```
(install
 (section lib)
 (files (mirage-xen.pc as ../pkgconfig/mirage-xen.pc)))
```
That file ends up in `_build/install/default/lib/pkgconfig/mirage-xen.pc` which seems like it's escaping the package's install directory by design. @rgrinberg is that right? That issue was made in 2019. Do we know if mirage still needs this feature?

### CODE REVIEW

**Comment by Alizter** [comment]
Can we get around with this constraint using symlinks?

**Comment by gridbugs** [comment]
> Can we get around with this constraint using symlinks?

How would that work? Could we allow files outside their package directory provided that they are symlinks?

**Comment by Leonidas-from-XIV** [comment]
> How would that work? Could we allow files outside their package directory provided that they are symlinks?

I guess if you symlink `..` to `foo` in your source tree you could write `foo/bar` and have `bar` end up in the parent folder (not that I tested it). Also not sure what happens if you write `foo/bar` to an `.install` file, whether OPAM will create symlink directories and write through symlink directories. It all feels very "exploity".

**Comment by rgrinberg** [comment]
@gridbugs the fix looks right to me. However, it has to be guarded behind a dune language version. Before we remove the old behavior, we need to go through a deprecation cycle as well.

> That file ends up in _build/install/default/lib/pkgconfig/mirage-xen.pc which seems like it's escaping the package's install directory by design. @rgrinberg is that right? That issue was made in 2019. Do we know if mirage still needs this feature?

I don't know about mirage, but installing pkg-config metadata is something that I'm sure many people are going to need.

@kit-ty-kate what is the recommended way to install such files?

**Comment by kit-ty-kate** [comment]
> @kit-ty-kate what is the recommended way to install such files?

I think using [lib_root](https://opam.ocaml.org/doc/Manual.html#installfield-lib_root) instead of `lib` would be the recommended way to do this.

**Comment by gridbugs** [comment]
Ok so for this PR make it so that the current behaviour is maintained prior to the the next release of dune after which we'll start printing a deprecated message (but it won't be an error).

Should I mention the deprecation in the docs or should the docs just say that dst paths may not begin with ".."?

**Comment by rgrinberg** [comment]
> this is in the test for installing source trees and it looks like this test is deliberately escaping its install directory (everything should be under .../default/doc/mypkg/). Is that intended?

The test is indeed intentionally testing this behavior. For `source_trees`, we can just disable escaping the section root out right since this feature hasn't been introduced.

**Comment by rgrinberg** [comment]
> Should I mention the deprecation in the docs or should the docs just say that dst paths may not begin with ".."?

I think you should mention it. It's useful information for the user, so of course it can only help if it's in the manual.

I wish we more systematic about mentioning versions in the manual as well. This would save users a bunch of time.

**Comment by gridbugs** [comment]
Starting a dst with `..` is now an error when installing source trees and a warning when installing files and dirs. I've added a bunch of tests of this behaviour and also of the `with_prefix` feature when using globs in the install stanza. I added docs for both features as well and also ended up adding more docs and tests for installing source trees to clarify some edge cases.

Added @christinerose to review the documentation changes.

**Comment by Alizter** [comment]
i'm still a big confused about the motivation. Why do we want to restrict inatall directories? That seems like we will be more opinionated than we need to be.

File: src/dune_rules/file_binding.ml [hunk 1]
```diff
@@ -4,26 +4,77 @@ open Memo.O
...
+  dune_syntax
+  =
+  if relative_path_starts_with_parent dst
+  then (
+    match relative_dst_path_starts_with_parent_error_when with
+    | `Deprecation_warning_from_3_11 ->
+      let open Syntax.Version.Infix in
+      if dune_syntax >= (3, 11)
+      then
+        User_warning.emit
```
Comment (rgrinberg) [inline src/dune_rules/file_binding.ml:37]: Warnings should be disabled in vendored projects.

File: src/dune_rules/install_entry.ml [hunk 2]
```diff
@@ -13,11 +13,46 @@ let expand_str_with_check_for_local_path ~expand_str sw =
...
+  let decode =
+    let open Dune_lang.Decoder in
+    let install_glob_version_check = Dune_lang.Syntax.since Stanza.syntax (3, 6) in
+    let install_glob_with_prefix_version_check =
+      Dune_lang.Syntax.since Stanza.syntax (3, 11)
+    in
+    let decode_args ~recursive =
+      let decode_without_prefix =
+        String_with_vars.decode
+        >>| fun glob ->
```
Comment (rgrinberg) [inline src/dune_rules/install_entry.ml:31]: nitpick: this style is never an improvement over:

```
let+ glob = String_with_vars.decode in
...
```

The only time we should use `>>|` or `>>=` is whenever we get to do something point-free. Otherwise, it's just making us read `let` bindings in reverse order.

File: src/dune_rules/install_entry.ml [hunk 3]
```diff
@@ -13,11 +13,46 @@ let expand_str_with_check_for_local_path ~expand_str sw =
...
+        { glob_files = { Dep_conf.Glob_files.glob; recursive }; prefix = None }
+      in
+      let decode_with_prefix =
+        enter
+          (let* () = install_glob_with_prefix_version_check in
+           let* glob = String_with_vars.decode in
+           let* _ = keyword "with_prefix" in
```
Comment (rgrinberg) [inline src/dune_rules/install_entry.ml:38]: Might as well make it `let* () =` here.

File: src/dune_rules/install_entry.ml [hunk 4]
```diff
@@ -26,36 +61,55 @@ module File = struct
...
+        List.map (Glob_files_expand.Expanded.matches glob_expanded) ~f:(fun path ->
           let src = glob_loc, path in
-          File_binding.Unexpanded.make ~src ~dst:src)
+          let dst =
+            match prefix_loc_opt with
+            | None -> src
+            | Some (prefix, prefix_loc) ->
+              let path_without_prefix =
+                String.drop_prefix path ~prefix:glob_prefix
+                |> function
```
Comment (rgrinberg) [inline src/dune_rules/install_entry.ml:93]: This is the same as writing:

```
match String.drop_prefix path ~prefix:glob_prefix with
```

I'm not against using `|>`, but there's no need to favor it over basic constructs like `match`.

**Comment by rgrinberg** [comment]
Looks good to me with a few minor suggestions remaining.

I don't know if this possible at this point, but I would have liked `with_prefix` to come in a subsequent PR.

Also, I think you need to add a few test cases to make sure `with_prefix` is being validated correctly. For example, absolute paths shouldn't be allowed there.

@Leonidas-from-XIV could you give this one some review as well?

**Comment by gridbugs** [comment]
> While I am not sure this is a good idea from a compatibility point of view (I guess this could prevent a number of packages upgrading to dune-lang 3.11 since we take away something without providing a clear way to migrate to) the code is fine. Some minor comments attached.

Regarding compatibility I can take a look through the monorepo benchmark duniverse to see how many packages take advantage of the `as` keyword to escape their install directory. We know that some packages do this to add pkg-config entries but it would be interesting to see if there are any other use cases that will be affected by this change.

For pkg-config the migration is to use the `lib_root` section. Where's the right place to add migration notes? Should it go in the changelog?

File: src/dune_rules/file_binding.ml (hunk 1 above)
Comment (gridbugs) [reply src/dune_rules/file_binding.ml:37]: I don't understand this comment. What are you suggesting I change here?

File: src/dune_rules/install_entry.ml [hunk 5]
```diff
@@ -66,40 +120,68 @@ module File = struct
     ;;
   end
 
-  type t = Without_include.t Recursive_include.t
+  type t =
+    { binding : Without_include.t Recursive_include.t
```
[the remark this answers is not in the document]
Comment (gridbugs) [reply src/dune_rules/install_entry.ml:124]: I renamed it to `entry` as it represents a single entry in the install stanza.

**Comment by gridbugs** [comment]
> I don't know if this possible at this point, but I would have liked `with_prefix` to come in a subsequent PR.

I've reduced this PR to just the deprecation of dsts starting with "..". I'll move the rest to a new PR.

**Comment by gridbugs** [comment]
I moved the `with_prefix` change to here: https://github.com/ocaml/dune/pull/8416

**Review by Leonidas-from-XIV — APPROVED** [review]

**Comment by rgrinberg** [comment]
@gridbugs I rebased the PR. I think it's still showing up warnings in vendored directories. It should be silenced there.

File: src/dune_rules/file_binding.ml (hunk 1 above)
Comment (rgrinberg) [reply src/dune_rules/file_binding.ml:37]: In a project that's under a `(vendored_dirs ..)`, we omit all the warnings. This is so that projects that emit projects used by opam-monorepo (or vendored otherwise), do not spam the main projects with warnings.

**Comment by gridbugs** [comment]
@rgrinberg if I'm correctly understanding you regarding vendored directories the problem would only happen if dune tried to build the install manifest for a vendored project. I don't know if it's possible for this to happen. I added a test to the end of `test/blackbox-tests/test-cases/start-install-dst-with-parent-error.t` that I thought might trigger this but it didn't work. Any idea how make dune evaluate the install stanza inside a vendored project?

**Comment by rgrinberg** [comment]
You're right of course. We disable install rules for vendored packages anyway. Anyway, it's just something to keep in mind in general then.
